[dv/doc] Adding the first update to coverage methodology (#3114)
lowRISC/opentitan: doc/ug/dv_methodology/index.md

## Coverage Collection

> Collecting, analyzing, and reporting coverage with waivers is a requirement to assert 'verification complet'.

Reviewer: Sorry I didn't notice this last night: "complete" has lost its trailing "e".

Author: wierd.. but fixed.. thank you

Any gaps in our measured coverage need to be understood and either waived (no need to cover) or closed by additional testing.
The end goal is to achieve 100% coverage across all applicable coverage metrics.
This process is known as "coverage closure", and is done in close collaboration with the designer(s).
Coverage collected from all tests run as a part of the regression is merged into a database for analysis.
Our primary tool of choice for our coverage closure needs is Synopsys VCS & Verdi.
However, the use of other simulators is welcome.

Why do we need coverage?

The common answer is to flush out bugs in the design.

> This is not accurate enough. Making sure there are no bugs in a design, While it is important to make sure there are no bugs in a design, this is not sufficient.

Author: thank you

One must also make sure the design works as intended.
That is, it must provide all the functionality specified in the design specification.

> So a more precise answer for why we need coverage is to flush out flaws in the design.
> These flaws can be either design bugs or deficiencies in the design with respect to the specification.

Reviewer 2: ```suggestion
So, we need coverage to flush out not only the bugs in the design but also deficiencies in the specification. 
```

Author: @sriyerg  I think I prefer my version.
but if you insist I don't feel that strongly about it.?

Reviewer: I suppose Sri's suggestion has the advantage that it doesn't call a deficiency in the specification a flaw in the design.

Reviewer 2: If we can convey the same idea in a more concise way, I'd prefer that. I will leave this upto you. I am fine either way.

  It is very difficult, and not particularly useful to achieve 100% toggle coverage across a design.
  Instead, we focus on closing toggle coverage only on the IO ports of the DUT and IO ports of pre-verified IPs within the DUT.
* **FSM state Coverage**: This metric measures which finite state machine states were executed during a simulation.
* **FSM transition Coverage**: This metric measures which arcs were simulated for each finite state machine in the design.
* **Conditional Coverage**: This metric tracks all combinations of conditional expressions simulated.
* **Branch Coverage**: This metric is similar to line coverage, but not quite the same.
  It tracks the flow of simulation (e.g. if/else blocks) as well as conditional expressions.
  Note that un-hit FSM state/transition coverage almost always shows up as un-hit branch coverage as well.

Code coverage is sometimes referred to as implicit coverage as it is generated based on the code and takes no additional effort to implement.

### Functional Coverage

Unlike code coverage, functional coverage requires the designer and/or DV engineer to write additional cover points and covergroups.

> For this reason functional coverage is sometimes referred to as explicit coverage.

Reviewer: I'm ambivalent about the comma. I'm fine either way with the quotes but, if we add them, please can they go around both words ('... "explicit coverage".')? The point is that "explicit coverage" is then a term of art, but "explicit" on its own isn't.

Incidentally, I always imagine that "explicit coverage" is found on the top shelf in seedy DV parlours, but maybe that's just me :-)

> For this reason functional coverage is sometimes referred to as explicit coverage.

Author: I removed the quotes I don't think they are needed.

> These are more complex constructs that capture whether signals (that reflect the current state of the design) have met a set or a sequence of values that are interesting (often called corner cases).

Reviewer: I realise that you've already reworded this once after a comment from me but, looking at the result, I think it can be shortened to "have met an interesting set or a sequence of values (often called corner cases)". (Sorry!)

Author: you are forgiven- and the text is updated

  This is called a coverage cross.
* **Cover Property coverage**: This observes events or series of events.
  As an example think of the TL UL register bus used in OpenTitan.
  A cover property cover point could be the handshaking between valid and ready.
  There is a big overlap in what can be covered by SVA and cover properties.
  Suggestions on when to use which form of coverage are presented later in this document.

#### Do we need both types of coverage

Reaching a 100% code coverage does not tell you the whole story.
Think of the simple example of an AND gate with inputs A, B, and output X.
To get to 100 percent code coverage only two different input combinations are needed.
00 and 11, these two will produce all possible outcomes of X.

![single AND gate](dv_method_single_gate.svg)

The coverage will indicate that all code was exercised.
But we do not know that our design works as intended.
All we know is that A, B, and X have been observed to take on both logic 0 and 1.
We could not say for certain that the design was in fact an AND gate it could just as easily be an OR gate.

> So we need functional coverage to tell us this.

Reviewer 2: Are we saying we want functional coverage to be implemented mandatorily regardless of how we are testing it?
Seems like a directed test (as opposed to constrained-random) that checks all combinations of A and B against the output Y can prove that it is indeed an AND gate without needing functional coverage to be instrumented. 

Functional coverage is no doubt, a concrete proof that we did verify a particular design functionality. However, at the chip level for example, we have almost exclusively all SW based directed tests. Schedule-wise it is a prohibitive to write functional coverage for all of those tests to prove that we did verify what we expected to verify. So we instead rely on 'testplan coverage' (all tests indicated in our testplan are passing in the regressions). Let me know your thoughts on this.

Reviewer: From my point of view, there are two questions:

1. Is this enough to sign off the chip?
2. Is this enough to be absolutely sure we've got it right?

If our signoff process isn't silly, we should definitely have 2 => 1, but it seems reasonable to tape out a chip when you're just pretty confident that it's correct (the exact value of "pretty" here depends on schedule pressure from the marketing department, your personal level of risk aversion, the company budget etc.!)

I think Rasmus is right in saying that these functional coverage points are required if you want to get 2. Otherwise, how are we "testing the tests"? If I'm trying to exercise something non-trivial, I can write a UVM sequence that I think should get me in a corner. I can even look at some waves to check I got it right. But what happens if the design is modified? Do I have to recheck the waves for each of my sequences? (clearly, this isn't going to happen!). I guess that functional coverage collection is a way to automate the "looks good to me" check that the DV engineer does while writing the tests in the first place.

Whether they are required for 1 is a judgement call, I guess. Personally, I'd try to do them, just to avoid having to worry near the tapeout deadline.

Author: @sriyerg @rswarbrick 
As Rupert says, I am trying to convey the message - if you don't have functional coverage how do you know you when you are done testing - how do you know the tests were good enough.
seems a bit arbitrary to me to say we are done when all of the tests we could think if making are passing.

I think we are hitting a cost vs timeline common issue.
In my experience it is always cheaper to go the extra mile and implement the functional coverage rather than relying upon directed tests as there are always corners that a missed.
and usually every respin means an additional million USD in cost and a 3-6 month schedule hit.

so unless you are depending on making the date of a specific shuttle run I would say - do functional testing.

there should be some value in chaining the IP environments.

Reviewer 2: You know when you are done testing through completion of the well reviewed testplan (i.e. testplan annotated regression result hitting 100%) and the integration code coverage closure. For SoC level integration testing (with already pre-verified IPs), this is often more than sufficient.

I am fine with the wording - I will just update the chip DV doc to say that we are waiving off the requirement to develop the functional coverage model.

The first thing functional coverage will tell us is if we observed all possible values on the inputs.
And by adding a cross between the inputs and the output it will tell us which gate we are looking at.

> Reaching 100 percent functional coverage is not enough either.

Author: fixed

> Remember functional coverage requires the designer to manually add coverage point into the design.

Reviewer 2: I am totally fine with you not agreeing with my comments, but I would at least let the original reviewer mark them as resolved, especially when they are not  actually resolved...

Going back to our AND gate, let us say we take two of these and OR the outputs of the two.

![multi gate](dv_method_multi_gate.svg)

> If we only add the cover points from our original example, these will still exercise the new output of our system and therefore result in reaching 100 percent functional coverage, but half of the design was not exercised.

Reviewer 2: nit: 100%

Author: fixed

This is called a coverage hole and code coverage would have indicated that a part of the code was never exercised.

> While functional coverage will tell you if your design is working correctly, code coverage will tell you if your testplan is incomplete or there are undocumented or unreachable features in the design.

Reviewer: Oh, sorry, I didn't notice this before. At the moment, this can be read as an either/or ("Hey, coverage! Please tell me what I got wrong: an incomplete test plan or undocumented features.").

Unfortunately, the syntax "can tell you if (A || B)" isn't allowed! You could turn it around like this, maybe? "..., code coverage will spot a hole if either your testplan is incomplete or if there are ..." (The second "if" is optional here, I think, but I reckon the sentence flows more nicely with it).

Author: updated

> As such they can be addressed accordingly through either updating the design specification and augmenting the testplan / written tests, or by optimizing the design to remove unreachable logic if possible.

Author: fixed

## Coverage Collection

Collecting, analyzing, and reporting coverage with waivers is a requirement to assert ‘verification complete’.
Any gaps in our measured coverage need to be understood and either waived (no need to cover) or closed by additional testing.
The end goal is to achieve 100% coverage across all applicable coverage metrics.
This process is known as “coverage closure”, and is done in close collaboration with the designer(s).
Coverage collected from all tests run as a part of the regression is merged into a database for analysis.
Our primary tool of choice for our coverage closure needs is Synopsys’ VCS & Verdi.

> However, the use of other tools/simulators are welcome.

Author: fixed

Why do we need coverage?

The common answer is to flush out bugs in the design.

> This is not accurate enough. Making sure there are no bugs in a design, while important it is not sufficient.

Author: fixed

> One must also make sure the design work as intended.

Author: sorry I thought I fixed that already. it is in there now

> That is providing all the functionality specified in the design specification.

Author: thank you, fixed

> So a more precise answer as to why we need coverage is to flush out flaws in the design.

Author: thank you. I was obviously failing :D

> Flaws being both design bugs and deficiencies in the design with respect to the specification.

Author: fixed

Another reason as to why we need coverage is to answer the seemingly simple but important question:
**When are we done testing?**

> Do we need 1, 10, or 100 tests and should they run 10,100, or 1000 regressions?

Author: fixed

Only coverage can answer this question for you.

There are two key types of coverage metrics: code coverage and functional coverage.
Both are important and are covered in more detail below.

> For this topic, we define ‘pre-verified sub-modules’ as IPs within the DUT that have already been (or are planned to be) verified to complete sign-off within individual test benches.

Author: @sriyerg 
I think this is your text, so I want to make sure you are ok with this change

### Code Coverage

Commercial simulators typically provide a way to extract coverage statistics from our regressions.
Tools automatically analyze the design to extract key structures such as lines, branches, FSMs, conditions, and IO toggle and provide them as different metrics to measure coverage against (to what extent is our stimulus through constrained random tests covering these structures).
These metrics are explained briefly below:

* **Line Coverage**: This metric measures which lines of SystemVerilog RTL code were executed during the simulation.
  This is probably the most intuitive metric to use.
  Note that `assign` statements are always listed as covered using this metric.

> * **Toggle Coverage**: This metric measures every logic bit to see if it transitions from 1 → 0 and 0 → 1.

Author: I think I copied this from a google doc hence the change. fixed

  Instead, we focus on closing toggle coverage only on the IO ports of the DUT and IO ports of pre-verified IPs within the DUT.
* **FSM state Coverage**: This metric measures which finite state machine states were executed during a simulation.
* **FSM transition Coverage**: This metric measures which arcs were simulated for each finite state machine in the design.
* **Conditional Coverage**: This metric tracks all combinations of conditional expressions simulated.
* **Branch Coverage**: This metric is similar to line coverage, but not quite the same.
  It tracks the flow of simulation (e.g. if/else blocks) as well as conditional expressions.
  Note that un-hit FSM state/transition coverage almost always shows up as un-hit branch coverage as well.

Code coverage is sometimes referred to as implicit coverage as it is generated based on the code and takes no additional effort to implement.

### Functional Coverage

Unlike code coverage, functional coverage requires the designer and/or DV engineer to write additional cover points and covergroups.
For this reason functional coverage is sometimes referred to as explicit coverage.

> These are more complex constructs that capture whether signals (that reflect the current state of the design) have met a set of values or a sequence of values that are interesting / corner cases.

Author: fixed

These constructs also allow us to capture whether multiple scenarios have occurred simultaneously through crosses.
This is also often referred to as cross coverage.
These constructs are typically encapsulated in a class and are sometimes referred to as the ‘functional coverage model’.
They are sampled in ‘reactive’ components of the testbench, such as monitors and/or the scoreboards.
They can also be embedded within the RTL to sample interesting scenarios through DV stimulus.

Here are the metrics used with a brief explanation:

* **Assertion Coverage using SVA**: This metric measures which assertions have been exercised in the course of the simulation.

> SVA coverage often implemented by Design engineer and is used to add cover procedures and variables not visible on output pins.

Author: thank you , fixed

  Note, an assertion precondition counts as a cover point.

> * **Covergroup Coverage**:Observes values on buses, registers and so on.

Author: absolutely

>   This can verify that a specific value or range of values was seen on a bus, And that no illegal values were seen.

Author: thank you

>   Simulators will throw an error if an illegal bin is filled.

Author: maybe just leave it at illegal values.. I think for now that is enough information

>   Could also be the intersection of two cover points.

Author: done

  One example could be observing the FIFO full signal and the FIFO write enable signal is asserted at the same time.

>   This is called a coverage cross and will give valuable information on whether this scenario was ever seen. 

Author: ok

* **Cover Property coverage**: This observes events or series of events.

>   As an example think of the TL UL register bus used.

Author: fixed

  A cover property cover point could be the handshaking between valid and ready.

>   There is a big overlap in what can be covered by SVA and Cover properties.

Author: thank you

>   Suggestions on when to use which form of coverage is presented later in this document.

Author: yes, there will be. thank you

#### Do we need both types of coverage

Reaching a 100% code coverage does not tell you the whole story.
Think of the simple example of an AND gate. 
Input A, B, and output X. to get to a 100 percent code coverage only two different input combinations are needed.

> 00 and 11, these two will produce all possible outcomes of X.

Author: I think the figure (not sure you looked at the pretty format)
will avoid anyone thinking about the wrong thing here.
what do you think?

![single AND gate](dv_method_single_gate.svg)

The coverage will indicate that all code was exercised.

> But We do not know that our design work as intended. 

Author: fixed

> All we know is that A, B, and X have been observed to take on both logic  0 and 1.

Author: fixed

> We could not say for certain that the design was in fact an AND gate could just as easily be an OR gate.

Author: do you want the colon in there?

> We could not say for certain that the design was in fact an AND gate could just as easily be an OR gate.

Reviewer: Yes, because you need something to separate the two clauses. Equally correct (but a bit stuffy looking) would be a semicolon. Or you could split the sentence in two.

> We could not say for certain that the design was in fact an AND gate could just as easily be an OR gate.

Author: ok fixed, not sure I like splitting them so added the colon

So we need functional coverage to tell us this.
The first thing functional coverage will tell us is if we observed all possible values on the inputs.

> And by Adding a cross between the inputs and the output it will tell us which gate we are looking at.

Author: fixed

Reaching 100 percent functional coverage is not enough either.
Remember functional coverage requires the designer to manually add coverage point into the design.
Going back to our AND gate, let us say we take two of these and OR the outputs of the two.

![multi gate](dv_method_multi_gate.svg)

If we only add the cover points from our original example.
These will still exercise the new output of our system and therefore result in reaching 100 percent functional coverage.

> But half of the design was not exercised.

Reviewer: I would suggest combining these three lines into a single sentence. (The first two need combining, and I think all three works better)

Author: I agree

This is called a coverage hole and code coverage would have indicated that a part of the code was never exercised.

While functional coverage will tell you if your design is working correctly.

> Code coverage will tell you if your testplan is incomplete or there are undocumented or unreachable features in the design.

Reviewer: Again, the sentences need combining, so "... is working correctly, code coverage will ..."

Author: done

## Coverage Collection

Collecting, analyzing, and reporting coverage with waivers is a requirement to assert 'verification complete'.
Any gaps in our measured coverage need to be understood and either waived (no need to cover) or closed by additional testing.
The end goal is to achieve 100% coverage across all applicable coverage metrics.
This process is known as "coverage closure", and is done in close collaboration with the designer(s).
Coverage collected from all tests run as a part of the regression is merged into a database for analysis.
Our primary tool of choice for our coverage closure needs is Synopsys VCS & Verdi.
However, the use of other simulators is welcome.

> Why do we need coverage?

Author: I removed the bold as Sri didn't like it as it looked like an additional heading.

> Why do we need coverage?

Reviewer 2: Yeah, this could be in bold. What I had suggested earlier was to not make it a level 4 heading.

> Why do we need coverage?

Author: ok done

The common answer is to flush out bugs in the design.
This is not accurate enough.

> Making sure there are no bugs in a design, While it is important to make sure there are no bugs in a design, this is not sufficient.

Author: oh yes good catch,

  Instead, we focus on closing toggle coverage only on the IO ports of the DUT and IO ports of pre-verified IPs within the DUT.
* **FSM state Coverage**: This metric measures which finite state machine states were executed during a simulation.
* **FSM transition Coverage**: This metric measures which arcs were simulated for each finite state machine in the design.
* **Conditional Coverage**: This metric tracks all combinations of conditional expressions simulated.
* **Branch Coverage**: This metric is similar to line coverage, but not quite the same.
  It tracks the flow of simulation (e.g. if/else blocks) as well as conditional expressions.
  Note that un-hit FSM state/transition coverage almost always shows up as un-hit branch coverage as well.

Code coverage is sometimes referred to as implicit coverage as it is generated based on the code and takes no additional effort to implement.

### Functional Coverage

Unlike code coverage, functional coverage requires the designer and/or DV engineer to write additional cover points and covergroups.
For this reason functional coverage is sometimes referred to as explicit coverage.

> These are more complex constructs that capture whether signals (that reflect the current state of the design) have met an interesting set or a sequence of values (often called corner cases)

Author: ty

These constructs also allow us to capture whether multiple scenarios have occurred simultaneously through crosses.
This is also often referred to as cross coverage.
These constructs are typically encapsulated in a class and are sometimes referred to as the 'functional coverage model'.
They are sampled in 'reactive' components of the testbench, such as monitors and/or the scoreboards.
They can also be embedded within the RTL to sample interesting scenarios through DV stimulus.

Here are the metrics used with a brief explanation:

* **Covergroup Coverage**: Observes values on buses, registers and so on.
  This can verify that a specific value or range of values was seen on a bus and that no illegal values were seen.
  Simulators can be set to throw an error if an illegal value is seen.

>   Another example of a covergroup could be the intersection of two cover points.

Author: thank you I like the rewording suggestion

  One example could be observing the FIFO full signal and the FIFO write enable signal is asserted at the same time.
  This is called a coverage cross.
* **Cover Property coverage**:
   * **Assertion coverage using SVA**

>    * **procedural code**

Author: they are just bullets and sub-bullets. no headings.
The Bold font was carried over from the original document.

> Most often Property coverage is implemented using Systemverilog Assertions (SVA).

Author: thank you

This observes events or series of events.
As an example think of the TL UL register bus used in OpenTitan.
A cover property cover point could be the handshaking between valid and ready.

> SVA also allows the Design engineer to add cover procedures and variables not visible on output pins.

Author: this a word was definitely missing there. thank you

> SVA also allows the Design engineer to add cover procedures and variables not visible on output pins.

Reviewer: I'm really sorry, but you've still left a capital in the middle of the sentence.

> SVA also allows the Design engineer to add cover procedures and variables not visible on output pins.

Author: I don't see it :S

> SVA also allows the Design engineer to add cover procedures and variables not visible on output pins.

Reviewer: Oops, I'm sorry! I think I meant to leave that note on a different line (314).

Note, an assertion precondition counts as a cover point.

#### Do we need both types of coverage

Reaching a 100% code coverage does not tell you the whole story.
Think of the simple example of an AND gate with inputs A, B, and output O.
To get to 100% code coverage only two different input combinations are needed.

> 00 and 11, these two will produce all possible outcomes of O.

Reviewer: The second sentence doesn't really work because it starts with a number (and there is a missing verb in the first clause). The simplest fix is probably: "To get to 100% code coverage only two different input combinations are needed: 00 and 11. These two combinations will produce all possible values of O."

Author: thank you, I was not aware that a sentence cannot start with numbers

Reviewer: I'm not sure if it's a hard-and-fast rule, but it looks a bit weird and I don't think I've ever seen it in print.

The coverage will indicate that all code was exercised.
But we do not know that our design works as intended.
All we know is that A, B, and O have been observed to take on both logic 0 and 1.
We could not say for certain that the design was in fact an AND gate: it could just as easily be an OR gate.
So we need functional coverage to tell us this.
The first thing functional coverage will tell us is if we observed all possible values on the inputs.
And by adding a cross between the inputs and the output it will tell us which gate we are looking at.

Reaching 100% functional coverage is not enough either.
Remember functional coverage requires the designer to manually add coverage point into the design.
Going back to our AND gate, let us say we take two of these and OR the outputs of the two.

![multi gate](dv_method_multi_gate.svg)

If we only add the cover points from our original example, these will still exercise the new output of our system and therefore result in reaching 100% functional coverage, but half of the design was not exercised.
This is called a coverage hole and code coverage would have indicated that a part of the code was never exercised.

> While functional coverage will tell you if your design is working correctly, code coverage will highlight if your testplan is incomplete, and if there are any undcovered/unreachable features in the design.

Reviewer: "undcovered". Also, I think this should be "or", not "and" (it spots if either thing happens, not only if both have happened).

Author: fixed

Author: did you make the same spelling mistake as me on purpose here :D

> Coverage holes can be addressed accordingly through either updating the design specification and augmenting the testplan / written tests, or by optimizing the design to remove unreachable logic if possible.

Reviewer: I'm not sure what "accordingly" means here and I think the sentence makes sense without it: just drop it, maybe?

Reviewer 2: I think by 'accordingly', @rasmus-madsen is referring to the 'under-covered' or 'unreachable' cases explained in the previous sentence. I will defer to your language expertise though.

Author: yep Sri is spot on.
I sometimes feel reading the markdown removes a bit of the flow adding newlines after a sentence.
reading the rendered document often helps.

also fixed "augmenting"
